File: apps/api/src/routes/shareSettings.ts
import { eq } from 'drizzle-orm';
import { Elysia, t } from 'elysia';
import { broadcast } from '../broadcast';
import { db } from '../db';
import { authMiddleware } from '../middleware/auth';
import { authorizeTrip } from '../middleware/authorize';
import { shareSettings } from '../schema';
import { sharePermissionSchema } from '../validators';

function generatePassphrase(): string {
  const chars = 'ABCDEFGHJKMNPQRSTUVWXYZ23456789'; // 紛らわしい文字を除外 (0/O, 1/I/L)
  const bytes = crypto.getRandomValues(new Uint8Array(6));
  return Array.from(bytes, (b) => chars[b % chars.length]).join('');
}

async function generateUniquePassphrase(): Promise<string> {
  for (let i = 0; i < 5; i++) {
    const passphrase = generatePassphrase();
    const [existing] = await db
      .select({ id: shareSettings.id })
      .from(shareSettings)
      .where(eq(shareSettings.shareToken, passphrase));
    if (!existing) return passphrase;
  }
  throw new Error('Failed to generate unique passphrase');
}

export const shareSettingsRoutes = new Elysia({
  prefix: '/trips/:tripId/share',
})
  .use(authMiddleware)

  // GET /trips/:tripId/share - Get share settings for a trip
  .get(
    '/',
    async ({ params, user, set }) => {
      const auth = await authorizeTrip(user.id, params.tripId, 'owner');
      if (!auth.authorized) {
        set.status = 403;
        return { error: 'Only the trip owner can manage share settings' };
      }

      const [setting] = await db
        .select()
        .from(shareSettings)
        .where(eq(shareSettings.tripId, params.tripId));
      if (!setting) {
        set.status = 404;
        return { error: 'Share settings not found' };
      }
      return setting;
    },
    {
      params: t.Object({
        tripId: t.String(),
      }),
    },
  )

  // POST /trips/:tripId/share - Create share settings with generated token
  .post(
    '/',
    async ({ params, body, user, set }) => {
      const auth = await authorizeTrip(user.id, params.tripId, 'owner');
      if (!auth.authorized) {
        set.status = 403;
        return { error: 'Only the trip owner can manage share settings' };
      }

      const shareToken = await generateUniquePassphrase();

      const [created] = await db
        .insert(shareSettings)
        .values({
          tripId: params.tripId,
          shareUrl: null,
          shareToken,
          permission: body.permission ?? 'view',
          createdBy: user.id,
        })
        .returning();
      broadcast(params.tripId, { type: 'INSERT', table: 'share_settings', record: created });
      set.status = 201;
      return created;
    },
    {
      params: t.Object({
        tripId: t.String(),
      }),
      body: t.Object({
        permission: t.Optional(sharePermissionSchema),
      }),
    },
  )

  // PUT /trips/:tripId/share - Update share settings
  .put(
    '/',
    async ({ params, body, user, set }) => {
      const auth = await authorizeTrip(user.id, params.tripId, 'owner');
      if (!auth.authorized) {
        set.status = 403;
        return { error: 'Only the trip owner can manage share settings' };
      }

      const updateData: Record<string, unknown> = {};
      if (body.permission !== undefined) updateData.permission = body.permission;
      if (body.isActive !== undefined) updateData.isActive = body.isActive;

      const [updated] = await db
        .update(shareSettings)
        .set(updateData)
        .where(eq(shareSettings.tripId, params.tripId))
        .returning();
      if (!updated) {
        set.status = 404;
        return { error: 'Share settings not found' };
      }
      broadcast(params.tripId, { type: 'UPDATE', table: 'share_settings', record: updated });
      return updated;
    },
    {
      params: t.Object({
        tripId: t.String(),
      }),
      body: t.Object({
        permission: t.Optional(sharePermissionSchema),
        isActive: t.Optional(t.Boolean()),
      }),
    },
  )

  // DELETE /trips/:tripId/share - Delete share settings
  .delete(
    '/',
    async ({ params, user, set }) => {
      const auth = await authorizeTrip(user.id, params.tripId, 'owner');
      if (!auth.authorized) {
        set.status = 403;
        return { error: 'Only the trip owner can manage share settings' };
      }

      const [deleted] = await db
        .delete(shareSettings)
        .where(eq(shareSettings.tripId, params.tripId))
        .returning();
      if (!deleted) {
        set.status = 404;
        return { error: 'Share settings not found' };
      }
      broadcast(params.tripId, { type: 'DELETE', table: 'share_settings', record: deleted });
      set.status = 204;
    },
    {
      params: t.Object({
        tripId: t.String(),
      }),
    },
  );

File: apps/api/internal/handler/share_settings.go
package handler

import (
	"crypto/rand"
	"errors"
	"math/big"
	"net/http"
	"strings"

	"github.com/jackc/pgx/v5"
	"github.com/labstack/echo/v4"

	db "github.com/shunto-ishiguro/trip-plan-app/apps/api/db/generated"
	"github.com/shunto-ishiguro/trip-plan-app/apps/api/internal/middleware"
	"github.com/shunto-ishiguro/trip-plan-app/apps/api/internal/ws"
)

const passphraseCharset = "ABCDEFGHJKMNPQRSTUVWXYZ23456789" // 30 chars, excludes O/I/L/0/1
const passphraseLength = 6
const maxPassphraseRetries = 5

type ShareSettingsHandler struct {
	queries *db.Queries
	hub     *ws.Hub
}

func NewShareSettingsHandler(queries *db.Queries, hub *ws.Hub) *ShareSettingsHandler {
	return &ShareSettingsHandler{queries: queries, hub: hub}
}

type createShareSettingsRequest struct {
	Permission string `json:"permission"`
}

type updateShareSettingsRequest struct {
	Permission *string `json:"permission,omitempty"`
	IsActive   *bool   `json:"isActive,omitempty"`
}

func generatePassphrase() (string, error) {
	b := make([]byte, passphraseLength)
	charsetLen := big.NewInt(int64(len(passphraseCharset)))
	for i := range b {
		idx, err := rand.Int(rand.Reader, charsetLen)
		if err != nil {
			return "", err
		}
		b[i] = passphraseCharset[idx.Int64()]
	}
	return string(b), nil
}

// GetShareSettings returns the share settings for a trip.
func (h *ShareSettingsHandler) GetShareSettings(c echo.Context) error {
	tripID := c.Param("tripId")
	ctx := c.Request().Context()

	s, err := h.queries.GetShareSettings(ctx, tripID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return echo.NewHTTPError(http.StatusNotFound, "share settings not found")
		}
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to query share settings")
	}

	return c.JSON(http.StatusOK, s)
}

// CreateShareSettings creates share settings for a trip with a unique passphrase token.
func (h *ShareSettingsHandler) CreateShareSettings(c echo.Context) error {
	tripID := c.Param("tripId")
	user := middleware.GetUser(c)
	ctx := c.Request().Context()

	var req createShareSettingsRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request body")
	}
	if req.Permission != "view" && req.Permission != "edit" {
		return echo.NewHTTPError(http.StatusBadRequest, "permission must be 'view' or 'edit'")
	}

	for attempt := 0; attempt < maxPassphraseRetries; attempt++ {
		token, err := generatePassphrase()
		if err != nil {
			return echo.NewHTTPError(http.StatusInternalServerError, "failed to generate share token")
		}

		s, err := h.queries.CreateShareSettings(ctx, db.CreateShareSettingsParams{
			TripID:     tripID,
			Permission: db.SharePermission(req.Permission),
			ShareToken: token,
			CreatedBy:  user.ID,
		})
		if err != nil {
			if strings.Contains(err.Error(), "unique") || strings.Contains(err.Error(), "duplicate") {
				continue
			}
			return echo.NewHTTPError(http.StatusInternalServerError, "failed to create share settings")
		}

		h.hub.Broadcast(tripID, ws.BroadcastEvent{
			Type:   "INSERT",
			Table:  "share_settings",
			Record: s,
		})

		return c.JSON(http.StatusCreated, s)
	}

	return echo.NewHTTPError(http.StatusInternalServerError, "failed to generate unique share token after retries")
}

// UpdateShareSettings updates permission and/or active status of share settings.
func (h *ShareSettingsHandler) UpdateShareSettings(c echo.Context) error {
	tripID := c.Param("tripId")
	ctx := c.Request().Context()

	var req updateShareSettingsRequest
	if err := c.Bind(&req); err != nil {
		return echo.NewHTTPError(http.StatusBadRequest, "invalid request body")
	}

	params := db.UpdateShareSettingsParams{
		TripID: tripID,
	}

	if req.Permission != nil {
		if *req.Permission != "view" && *req.Permission != "edit" {
			return echo.NewHTTPError(http.StatusBadRequest, "permission must be 'view' or 'edit'")
		}
		params.Permission = db.NullSharePermission{
			SharePermission: db.SharePermission(*req.Permission),
			Valid:           true,
		}
	}
	if req.IsActive != nil {
		params.IsActive = req.IsActive
	}

	s, err := h.queries.UpdateShareSettings(ctx, params)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return echo.NewHTTPError(http.StatusNotFound, "share settings not found")
		}
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to update share settings")
	}

	h.hub.Broadcast(tripID, ws.BroadcastEvent{
		Type:   "UPDATE",
		Table:  "share_settings",
		Record: s,
	})

	return c.JSON(http.StatusOK, s)
}

// DeleteShareSettings removes share settings for a trip.
func (h *ShareSettingsHandler) DeleteShareSettings(c echo.Context) error {
	tripID := c.Param("tripId")
	ctx := c.Request().Context()

	err := h.queries.DeleteShareSettings(ctx, tripID)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to delete share settings")
	}

	h.hub.Broadcast(tripID, ws.BroadcastEvent{
		Type:   "DELETE",
		Table:  "share_settings",
		Record: map[string]string{"tripId": tripID},
	})

	return c.NoContent(http.StatusNoContent)
}
